File: src/tools/actionTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { FptServerClient } from '../fptClient.js';
import { jobToMarkdown, Job } from '../utils/jobFormatter.js';
import { mcpText } from '../utils/mcpResponse.js';

function looksLikeJob(data: any): data is Job {
  return data && typeof data === 'object' && typeof data.state === 'string' && typeof data.action_name === 'string';
}

/**
 * Registers the generic dispatch tool that reaches every REST-exposed
 * action by name, mirroring `POST /actions/{name}` directly. This is the
 * fallback for anything not covered by a dedicated tool (fpt_trigger_build,
 * fpt_cancel_job, ...) — new server-side actions need no new tool here.
 * Use `fpt_describe_action` first to see the param schema for a given name.
 */
export function registerActionTools(server: McpServer, client: FptServerClient): void {
  server.registerTool(
    'fpt_invoke_action',
    {
      description:
        "Generic dispatch: invoke any REST-exposed action by name (see fpt_list_actions / fpt_describe_action). " +
        "Returns 202 with a job id for kind='job' actions, 200 with the result otherwise.",
      inputSchema: {
        name: z.string().describe("Action name, e.g. 'ci.build', 'ci.clean', 'cron.run'"),
        params: z.record(z.unknown()).optional().describe('Action-specific parameters, per fpt_describe_action'),
      },
    },
    async ({ name, params }) => {
      const result = await client.post<any>(`/actions/${encodeURIComponent(name)}`, params ?? {});
      if (looksLikeJob(result)) {
        return mcpText(jobToMarkdown(result));
      }
      return mcpText(JSON.stringify(result, null, 2));
    }
  );
}

File: lib/src/tools/action_tool.dart
import 'dart:convert';

import 'package:dart_mcp/server.dart';
import 'package:fpt_server_shared/fpt_server_shared.dart' as shared;

import '../fpt_client.dart';
import '../job_formatter.dart';
import '../mcp_response.dart';
import '../server.dart';

bool _looksLikeJob(Map<String, dynamic> data) =>
    data['state'] is String && data['action_name'] is String;

/// Registers the generic dispatch tool that reaches every REST-exposed
/// action by name, mirroring `POST /actions/{name}` directly. This is the
/// fallback for anything not covered by a dedicated tool (fpt_ci_build,
/// fpt_cancel_job, ...) — new server-side actions need no new tool here.
/// Use `fpt_describe_action` first to see the param schema for a given name.
void registerActionTools(FptMcpServer server, FptClient client) {
  server.registerTool(
    Tool(
      name: 'fpt_invoke_action',
      description:
          'Generic dispatch: invoke any REST-exposed action by name (see '
          "fpt_list_actions / fpt_describe_action). Returns 202 with a job "
          "id for kind='job' actions, 200 with the result otherwise.",
      inputSchema: Schema.object(
        properties: {
          'name': Schema.string(
            description: "Action name, e.g. 'ci.build', 'ci.clean', 'cron.run'",
          ),
          'params': Schema.object(
            description: 'Action-specific parameters, per fpt_describe_action',
          ),
        },
        required: ['name'],
      ),
    ),
    (request) async {
      final args = request.arguments!;
      final name = args['name'] as String;
      final params =
          (args['params'] as Map?)?.cast<String, Object?>() ?? const {};
      final result = await client.postJson(
        '/actions/${Uri.encodeComponent(name)}',
        params,
      );
      if (_looksLikeJob(result)) {
        return mcpText(jobToMarkdown(shared.Job.fromJson(result)));
      }
      return mcpText(const JsonEncoder.withIndent('  ').convert(result));
    },
  );
}
